### DISCUSSION TITLE
OAuth setup with Zitadel

### DISCUSSION THREAD

**Discussion by Aurel004** [mealie-recipes/mealie#3557] [OAuth Provider Example]
### Configuration Example

Let's configure mealie to work with Zitadel

Docs used:
- https://docs.mealie.io/documentation/getting-started/authentication/oidc/
- https://docs.mealie.io/documentation/getting-started/installation/backend-config/#openid-connect-oidc

1. Go to your zitadel console: https://zitadel.mydomain.com/ui/console

2. Go to _Projects_ and create a new one if needed, mine will be `server`

3. Click on _New_
![1](https://github.com/mealie-recipes/mealie/assets/20973812/a569157b-45b2-4669-b264-d2c82fc81df6)

4. _Name_ it `Mealie` and select _User Agent_
![image](https://github.com/mealie-recipes/mealie/assets/20973812/d82d8453-b93f-40a9-8300-745200bd6cf7)

5. Select _PKCE_
![image](https://github.com/mealie-recipes/mealie/assets/20973812/f048aacf-8af7-4fa0-b1b8-18d531fab0e2)

6. _Redirect URIs_ set the url of mealie and path _login_: `https://mealie.mydomain.com/login`
_Post Logout URIs_ set the url of mealie and path _login?direct=1_: `https://mealie.mydomain.com/login?direct=1`
![4](https://github.com/mealie-recipes/mealie/assets/20973812/bf478512-d20f-4659-84db-c5ba8efbd602)

7. Click _Create_ and copy the ClientId provided in Notepad

8. To configure the _OIDC_ADMIN_GROUP: Users that are in this group (within your IdP) will be seen as admin in Mealie._ you will need to configure _Roles_
    - In your project (_Step 2_), click on _General_, and copy to Notepad the _Resource Id_ shown on top
    - In your project (_Step 2_), click on _General_, and tick `Assert Roles on Authentication`
    - In your project (_Step 2_), click on _Roles_, _New_
![8](https://github.com/mealie-recipes/mealie/assets/20973812/9b3f4237-a7f2-44b0-9b9a-9849ac146d38)
    - Enter _Key_ `Admin` (this will be the name used for the group name in mealie), a display name and a group (this one is used for hierarchical groups in case you create other groups)
![image](https://github.com/mealie-recipes/mealie/assets/20973812/283a990c-b2e0-438e-bc59-b63d3df74228)
    - Next, go to _Authorizations_, click _New_, search your username (and/or people you want to be admin), click _Continue_ and select the created Role
    - Click on your newly created Mealie app in Zitadel (in your project, _General_), click on `Token Settings` and tick `User roles inside ID Token` to add the `Admin` role to the response for Mealie
    - Also tick `User Info inside ID Token` to give your _preferred_username_ and _email_ to Mealie

10. Go to your mealie docker-compose.yml file and add the following to _environment_:
`#SSO`
`OIDC_AUTH_ENABLED: true #Enable OIDC`
`OIDC_SIGNUP_ENABLED: false #Allow unknown users for mealie`
`OIDC_CONFIGURATION_URL: https://zitadel.mydomain.com/.well-known/openid-configuration`
`OIDC_CLIENT_ID: XXX@server #ClientId of Step 7`
`OIDC_AUTO_REDIRECT: false #Enable or disable auto-redirect to OIDC`
`OIDC_PROVIDER_NAME: Zitadel #Choose a name, it can be OAuth (default)`
`OIDC_REMEMBER_ME: true`
`OIDC_ADMIN_GROUP: Admin #The Key name in Step 8`
`OIDC_GROUPS_CLAIM: urn:zitadel:iam:org:project:{projectId}:roles #Replace {projectId} with the Resource Id copied in step 8, you can also put urn:zitadel:iam:org:project:roles instead but might be deprecated in the future`
`OIDC_USER_CLAIM: email #Default email, so your email in Zitadel must match existing email account in mealie (if OIDC_SIGNUP_ENABLED set to false), it can be changed to preferred_username`

11. Recreate your container and voilà, you're done

**Comment by shaqagain** [comment]
Is this setup still possible following v2.0? I'm having trouble setting up Mealie w/ Zitadel using User Agent/SPA (because there's no secret) and using Web. Mealie itself recognizes my OIDC configuration, but Zitadel does not even see the client ID.

**Reply by shaqagain** [reply]
I realized my mistake. I was under the impression the ".well-known/openid-configuration" portion of the OIDC_CONFIGURATION_URL was a placeholder, but its not. Working as expected after setting to https://sub.domain.com/.well-known/openid-configuration

Thank you for the guide!

**Comment by radiotaiso** [comment]
I was able to setup using web application and code authorization, same setup just remember to add the OIDC secret as well to the env vars, one important thing to setup is the `In your project (Step 2), click on General, and tick Assert Roles on Authentication` step, which is now called `Return user roles during authentication` Mealie expects those during login.

Thank you for the guide Aurel!
